### PR TITLE
Bring Your Own Resources

### PR DESCRIPTION
**Pull request by Yura52**
The branch adds a brand-new feature which allows to use custom servers for running containers.

### CODE REVIEW

File: etc/byor_config.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+c = get_config()
+load_subconfig('etc/base_config.py')
+load_subconfig('etc/github_auth.py')
+
+from dockerspawner import DockerSpawner
+c.DockerSpawner.hub_ip_connect = c.JupyterHub.hub_ip
```
Comment (anaderi) [inline etc/byor_config.py]: let's remove the last 2 lines. check that it doesn't break anything. if so, you can just load_subconfig('etc/local_config.py')

File: everware/_spawner_options_form.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    $('#byor_input').toggle(speed='normal');
+});
+</script>
+
+<div id='byor_input' style="display: none;">
+<p style="margin-bottom: 0px;">
+    For a successful run <a href="https://www.docker.com/" target="_black">Docker</a>
+    must be installed on your server.<br />
+    Enter ip and port of the Docker daemon running on your server.<br />
+    (Click <a href="https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option"
```
Comment (anaderi) [inline everware/_spawner_options_form.py]: should read:
Enter IP address and port of the Docker daemon running on your server.

File: everware/_spawner_options_form.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+SPAWNER_OPTIONS_FORM = """
+<div style="margin-bottom: 0px;">
```
Comment (anaderi) [inline everware/_spawner_options_form.py]: make separate html file in share/static/html and read it instead.

File: everware/spawner.py [hunk 4]
```diff
@@ -42,9 +44,41 @@ def __init__(self, **kwargs):
...
+            self._reset_byor()
+            return
+        byor_ip = self.user_options['byor_docker_ip']
+        byor_port = self.user_options['byor_docker_port']
+        # version='auto' causes a connection to the daemon.
+        # That's why the method must be a coroutine.
+        self._byor_client = docker.Client('{}:{}'.format(byor_ip, byor_port),
+                                          version='auto',
+                                          timeout=self.byor_timeout)
+        self.container_ip = byor_ip
```
Comment (anaderi) [inline everware/spawner.py]: move to ByorDockerSpawner class

File: etc/byor_config.py (hunk 1 above)
Comment (anaderi) [reply etc/byor_config.py]: just specify ByorDockerSpawner as spawner

File: everware/spawner.py [hunk 5]
```diff
@@ -331,7 +344,7 @@ def remove_old_container(self):
                 force=True
             )
         except APIError as e:
-            self.log.info("Can't erase container %s due to %s" % (self.container_name, e))
+            self.log.info("Can't erase container %s due to %s" % (self.container_name, e))
```
Comment (anaderi) [inline everware/spawner.py]: remove trailing space

**Review by anaderi — CHANGES_REQUESTED** [review]
please review

File: etc/byor_config.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+c = get_config()
+load_subconfig('etc/base_config.py')
+load_subconfig('etc/github_auth.py')
+
+c.JupyterHub.spawner_class = 'everware.ByorDockerSpawner'
+
+from os.path import join as pjoin
+with open(pjoin(c.JupyterHub.template_paths[0], '_byor_options_form.html')) as form:
+    c.ByorDockerSpawner.options_form = form.read()
```
Comment (anaderi) [inline etc/byor_config.py]: move to ByorDockerSpawner.__init__

**Review by anaderi — CHANGES_REQUESTED** [review]
and bump version as well

**Comment by yuvipanda** [comment]
This is awesome! Thank you for working on this!

However, I have some concerns about security & authentication. If you open up a docker daemon without any authentication to the internet, that pretty much provides attackers with root on that machine with no effort. There already is malware that looks for these kinds of vulnerabilities - https://duo.com/blog/over-18000-redis-instances-targeted-by-fake-ransomware for one that attacked redis, http://www.securityweek.com/elasticsearch-servers-latest-target-ransom-attacks for elasticsearch, https://nakedsecurity.sophos.com/2017/01/11/thousands-of-mongodb-databases-compromised-and-held-to-ransom/ for mongodb, etc. Since docker allows much more arbitrary compute execution than the others did (except for elasticsearch - there were attacks that used ES to spawn a VM!), it is also potentially used as command & control center for malware / ransomware attacks (unpatched wordpress instances were used in cyberattacks related to the US '16 Election, for example https://www.wordfence.com/blog/2017/02/russia-election-hack-worpress-used/). 

Docker has TLS support for doing client authentication (https://docs.docker.com/engine/security/https/). This will require users to provide everware a certificate as well as the docker daemon URL - the client certificate can then be kinda used like a password (@betatim's idea!). This would be a huge improvement over plain unauthenticated docker endpoints. I hope you would consider implementing it, and more importantly *requiring* it - since otherwise people would just probably skip it.

Thanks!

**Comment by anaderi** [comment]
Thanks Yuvi. It was just the first step. Tls part is on its way. Stay tuned!

**Comment by yuvipanda** [comment]
tyvm! This is great :D

**Pull request merged**